### ISSUE TITLE
keras转ncnn后，执行结果异常

### ISSUE THREAD

**Issue by lhxy1991** [Tencent/ncnn#3040]
使用MarsTechHAN的keras2ncnn转换一个简单的图像分类（8个类别），转换成ncnn模型后，按照正常的使用方法模型输出结果8个输出值都是0.125。
训练模型代码为：

> model = Sequential()

# add Con2D layers
model.add(Conv2D(filters=32, kernel_size=(3, 3), activation='relu', input_shape=(48, 48, 3)))
model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))

model.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu'))
model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))

model.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu'))
model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))

model.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu'))
model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))

# flatten
model.add(Flatten())

# dropOut layer
model.add(Dropout(0.2))

# add one simple layer for classification
model.add(Dense(units=512, activation='relu'))

# add output layer
model.add(Dense(units=8, activation='softmax'))

# compile model
model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['acc'])

# print model info
model.summary()
json_str = model.to_json()
print(json_str)
# fit_generator to fill in the dataset
history = model.fit_generator(
    generator=train_data_generator,
    steps_per_epoch=1000,
    epochs=100,
    validation_data=validation_data_generator,
    validation_steps=1000)

训练结果为
[traffic_signs.zip](https://github.com/Tencent/ncnn/files/6731751/traffic_signs.zip)
keras2ncnn可正常转换，但是转换后的param文件输入端的值不是48 48 3，是-1 -1 -1。
@nihui 麻烦帮忙看一下

**Comment by nihui** [comment]
param 直接用编辑器打开，Input 后面的参数删掉就好啦

**Comment by lhxy1991** [comment]
试过了，还是一样的结果。改成48 48 0 也是输出全是0.125。
keras使用的是NHWC,转换为ncnn后存储方式是否会发生改变呢？ ncnn::Mat 默认的是NCHW。使用from_pixels_resize() 和
// cv::Mat a(h, w, CV_32FC3); ncnn::Mat in_pack3(a.cols, a.rows, 1, (void*)a.data, (size_t)4u * 3, 3); ncnn::Mat in; ncnn::convert_packing(in_pack3, in, 1);（这种是否保留原格式NHWC?）两种方式，结果依旧都是0.125.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Tencent/ncnn"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年6月29日(星期二) 下午4:57
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Tencent/ncnn] keras转ncnn后，执行结果异常 (#3040)





 
param 直接用编辑器打开，Input 后面的参数删掉就好啦

**Comment by lhxy1991** [comment]
更改训练模型使用NCHW格式，转ncnn以后，训练结果依旧是0.12左右。




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "Tencent/ncnn"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2021年6月29日(星期二) 下午4:57
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [Tencent/ncnn] keras转ncnn后，执行结果异常 (#3040)





 
param 直接用编辑器打开，Input 后面的参数删掉就好啦

**Comment by superbayes** [comment]
> 使用MarsTechHAN的keras2ncnn转换一个简单的图像分类（8个类别），转换成ncnn模型后，按照正常的使用方法模型输出结果8个输出值都是0.125。 训练模型代码为：
> 
> > model = Sequential()
> 
> # add Con2D layers
> model.add(Conv2D(filters=32, kernel_size=(3, 3), activation='relu', input_shape=(48, 48, 3))) model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))
> 
> model.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu')) model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))
> 
> model.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu')) model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))
> 
> model.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu')) model.add(MaxPool2D(pool_size=(2, 2), padding='valid'))
> 
> # flatten
> model.add(Flatten())
> 
> # dropOut layer
> model.add(Dropout(0.2))
> 
> # add one simple layer for classification
> model.add(Dense(units=512, activation='relu'))
> 
> # add output layer
> model.add(Dense(units=8, activation='softmax'))
> 
> # compile model
> model.compile(loss='categorical_crossentropy', optimizer='rmsprop', metrics=['acc'])
> 
> # print model info
> model.summary() json_str = model.to_json() print(json_str)
> 
> # fit_generator to fill in the dataset
> history = model.fit_generator( generator=train_data_generator, steps_per_epoch=1000, epochs=100, validation_data=validation_data_generator, validation_steps=1000)
> 
> 训练结果为 [traffic_signs.zip](https://github.com/Tencent/ncnn/files/6731751/traffic_signs.zip) keras2ncnn可正常转换，但是转换后的param文件输入端的值不是48 48 3，是-1 -1 -1。 @nihui 麻烦帮忙看一下



> 
我认为问题可能出在tensor的shape发生变化,导致最终预测结果异常.
keras默认接受的tensor shape=NHWC, 但ncnn在将数据经过substract_mean_normalize函数处理后,tensor shape=NCHW
你可以试试如此处理:
![image](https://user-images.githubusercontent.com/56495942/144983594-0262ad63-ec1f-48ce-a61f-18a9840f0ff0.png)
